### PR TITLE
Add "remote" option for local dev setup

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/area dev-productivity
/area testing
/kind enhancement
/priority normal

**What this PR does / why we need it**:
Adds a new "remote" option for local dev setup. With this option, an existing remote Kubernetes cluster (e.g. a Gardener shoot on GCP) can be used as a garden cluster. This is needed to test e.g. shooted seeds with their own gardenlet, since a remotely running gardenlet can't connect to a locally running cluster.
Uses https://github.com/mvladev/quic-reverse-http-tunnel for UDP tunnelling. Big kudos to @mvladev for this!

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
The intended workflow is as follows:
```
# Bootstrap Remote garden
export KUBECONFIG=<kubeconfig>
make remote-garden-up
make dev-setup
make start-apiserver
...

# Cleanup remote garden
make remote-garden-down
```

**Release note**:
```improvement operator
NONE
```

### CODE REVIEW

**Comment by mvladev** [comment]
Certificates should be generated for every single `make remote-garden-up` - the authentication is happening via client certificates.

**Comment by stoyanr** [comment]
> Certificates should be generated for every single `make remote-garden-up` - the authentication is happening via client certificates.

@mvladev Yes, I will improve this.

**Comment by mvladev** [comment]
I've enabled github container registry actions for that repository. Now everything is and published automatically.

- `ghcr.io/mvladev/quic-reverse-http-tunnel/quic-server:v0.1.2`
- `ghcr.io/mvladev/quic-reverse-http-tunnel/quic-client-tcp:v0.1.2`

**Comment by stoyanr** [comment]
@mvladev Added generation of the tunnel certificates as suggested. The `gardener-apiserver` certificates that are only used locally are still hardcoded (similarly to many other certificates that are only used locally by the dev setup).

To avoid mistakes and reliably identify the cluster as "remote garden", I also added a check that the `KUBECONFIG` points to a cluster with a `garden` namespace labeled with `gardener.cloud/purpose=remote-garden`. The user must create the `garden` namespace and label it manually before running `make remote-garden-up`.

I consider this PR ready for review now.

File: hack/local-development/remote-garden/cleanup [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
```
Comment (rfranzke) [inline hack/local-development/remote-garden/cleanup:1]: license header

File: hack/local-development/remote-garden/open-gardener-tunnels [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+source $(dirname "${0}")/../common/helpers
+
+QUIC_SERVER_IMAGE=ghcr.io/mvladev/quic-reverse-http-tunnel/quic-server:v0.1.2
+QUIC_CLIENT_IMAGE=ghcr.io/mvladev/quic-reverse-http-tunnel/quic-client-tcp:v0.1.2
```
Comment (rfranzke) [inline hack/local-development/remote-garden/open-gardener-tunnels:20]: Should these images rather be moved into the gardener org/registry?

**Review by rfranzke — COMMENTED** [review]
Can you also enhance https://github.com/gardener/gardener/blob/master/docs/development/local_setup.md with a description of this kind of setup, please?

File: hack/local-development/remote-garden/open-gardener-tunnels (hunk 2 above)
Comment (mvladev) [reply hack/local-development/remote-garden/open-gardener-tunnels:20]: Well the images are build with this https://github.com/mvladev/quic-reverse-http-tunnel/blob/master/.github/workflows/cd_workflow.yaml and stored in the github container registry. Those would be used for local development and are not fetched all the time.

I guess they can be added to `images.yaml` and be mirrored?

Comment (rfranzke) [reply hack/local-development/remote-garden/open-gardener-tunnels:20]: The mirroring is only happening for productive images.

File: hack/local-development/remote-garden/cleanup (hunk 1 above)
Comment (stoyanr) [reply hack/local-development/remote-garden/cleanup:1]: Done

File: hack/local-development/remote-garden/open-gardener-tunnels (hunk 2 above)
Comment (stoyanr) [reply hack/local-development/remote-garden/open-gardener-tunnels:20]: As discussed, let's leave the images as they are for now.

**Comment by stoyanr** [comment]
@rfranzke @mvladev I addressed all feedback and added some documentation to https://github.com/gardener/gardener/blob/master/docs/development/local_setup.md as requested.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**